### DISCUSSION TITLE
Problems with TypeScript example: your_fungible_asset.ts

### DISCUSSION THREAD

**Discussion by ganymedio** [aptos-labs/aptos-developer-discussions#309] [Questions] [answered]
TL/DR: A couple examples of me trying to mint an FA and it not transferring. The issue appears to be related to that the store owner declined to allow ungated transfer. Can anyone advise how to solve that and allow mint deposits to work? 

<img width="586" alt="image" src="https://github.com/user-attachments/assets/7e460cb9-efaa-4271-b987-91cd6b717dc9">

Seems like I need to add `allow_ungated_transfer` somewhere. 

I see an object holds an `ObjectCore` resource:

```
struct ObjectCore has key {
    guid_creation_num: u64,
    owner: address,
    allow_ungated_transfer: bool,
    transfer_events: event::EventHandle<TransferEvent>,
}
```
just not clear how to integrate that into my module to allow ungated transfer. I don't see `ObjectCore` in the `fungible_asset` module at so not sure where to modify that parameter for a fungible asset.

I tried generating a transfer ref and then `object::enable_ungated_transfer(&transfer_ref);` and got back the error

```
Given: '(aptos_framework=0x1)::fungible_asset::TransferRef'

Expected: '(aptos_framework=0x1)::object::TransferRef'
```

Here's the `generate_transfer_ref` code from `fungible_asset.move`:

```
    public fun generate_transfer_ref(constructor_ref: &ConstructorRef): TransferRef {
        let metadata = object::object_from_constructor_ref<Metadata>(constructor_ref);
        TransferRef { metadata }
    }
```

I'm not quite understanding the connection between `metadata` as an object, and the two different `TransferRef` types... not sure if there's a way to convert between them somehow? Or another solution based on how FAs work? 

Before getting into my FACoin module, I tried the `your_fungible_asset` example in the Aptos TS SDK repo and got similar failure to transfer.

Expected output shown in [docs](https://aptos.dev/en/build/guides/first-fungible-asset):

```
===Publishing FAcoin package===
 
Transaction hash: 0x90fbe811171dde1ffad9157314ce0c4f6070fd5c1095a0e18a0b5634d10d7f48
metadata address: 0x77503715cb75fcc276b4e7236210fb0bcac7b510f6428233b65468b1cd3d708b
All the balances in this example refer to balance in primary fungible stores of each account.
Alice's initial FACoin balance: 0.
Bob's initial FACoin balance: 0.
Charlie's initial balance: 0.
Alice mints Charlie 100 coins.
Charlie's updated FACoin primary fungible store balance: 100.
Alice freezes Bob's account.
Alice as the admin forcefully transfers the newly minted coins of Charlie to Bob ignoring that Bob's account is frozen.
Bob's updated FACoin balance: 100.
Alice unfreezes Bob's account.
Alice burns 50 coins from Bob.
Bob's updated FACoin balance: 50.
Bob transfers 10 coins to Alice as the owner.
Alice's updated FACoin balance: 10.
Bob's updated FACoin balance: 40.
```

My output when I followed the directions in the docs:

```
===Publishing FACoin package===
Transaction hash: 0x4022765b50dd64b46561e179133c918e43cfdd245fc6265c27b8161548a884b1
metadata address: 0xd77d25faf07125fd7ccca1bbc671de762f2db4069924f1875b621c914c037aa0
All the balances in this exmaple refer to balance in primary fungible stores of each account.
Alice's initial FACoin balance: 0.
Bob's initial FACoin balance: 0.
Charlie's initial balance: 0.
Alice mints Charlie 100 coins.
Charlie's updated FACoin primary fungible store balance: 0.
Alice freezes Bob's account.
Alice as the admin forcefully transfers the newly minted coins of Charlie to Bob ignoring that Bob's account is frozen.
Bob's updated FACoin balance: 0.
Alice unfreezes Bob's account.
Alice burns 50 coins from Bob.
Bob's updated FACoin balance: 0.
Bob transfers 10 coins to Alice as the owner.
Alice's updated FACoin balance: 0.
Bob's updated FACoin balance: 0.
done.
```

So apparently this may be a more general issue, if the TS test isn't completing the transfers either?

I created a package with the FACoin example (code below for context). I published it on devnet under account `0xd2725edaa32bcd12a21efe8e0d7e0e7e029791b81814901196a74bd6dacdd07b`.

Then I ran

```
aptos move run --function-id 0xd2725edaa32bcd12a21efe8e0d7e0e7e029791b81814901196a74bd6dacdd07b::fa_coin::mint  --args address:0xb118a531aa5efb194f5e322aeee8e9d1e98c6220936d1cee90e1480911d51d64 u64:5
```

Result was 

```
{
  "Result": {
    "transaction_hash": "0xdd99861bf1887bdf9f400a00699856553dd7f5b4758b6a5330aa8b4d824a0702",
    "gas_used": 543,
    "gas_unit_price": 100,
    "sender": "d2725edaa32bcd12a21efe8e0d7e0e7e029791b81814901196a74bd6dacdd07b",
    "sequence_number": 1,
    "success": true,
    "timestamp_us": 1720851436040927,
    "version": 17613188,
    "vm_status": "Executed successfully"
  }
}
```

When I look up the txn hash `0xdd99861bf1887bdf9f400a00699856553dd7f5b4758b6a5330aa8b4d824a0702` in Aptos devnet explorer, I don't see any movement of `FACoin` nor do I see any `FACoin` under the `0xb118...` address.

I do see an event happened, a deposit to store: 

```
{
amount:"5"
store:"0x578e65b0f5a9d74a4a0515202437fbb58605201efd105bfe9d5618692afd6280"
}
```

but the `0xb118...` address only appears to have a store for `AptosCoin` from when I funded it with the faucet.

I tried it first without and then with the overrides commented out. Result appears to have been the same. 

Anyone else able to repro this and/or see whether I'm missing some step in the mint process? Thanks for any tips on how to get the FA to transfer successfully in these examples. The TS one is especially odd because I believe I followed the directions in the  docs but still got a failure to transfer.

**Comment by kshitijc1506** [comment]
Stores are objects. Every object has its associated owner. You can see that the deposit actually did work successfully [here](https://explorer.aptoslabs.com/object/0x578e65b0f5a9d74a4a0515202437fbb58605201efd105bfe9d5618692afd6280/resources?network=devnet). 

In the first scenario, it looks like you're trying to transfer the store itself. This is different from minting the assets.

**Reply by ganymedio** [reply]
Thanks, I wonder why the TS example didn't work as expected though. Shouldn't nonzero FACoin balances of Alice, Bob and Charlie shown up in the output?

**Reply by kshitijc1506** [reply, marked as answer]
I double checked the code and found this: 

```typescript
console.log(`Charlie's initial balance: ${await await getFaBalance(charlie, metadataAddress)}.`);
``` 

There are two `await`s. Can you fix that and try? It solved it for me. Also, make sure that your dependencies are all up to date (installation instructions in the [docs](https://aptos.dev/en/build/guides/first-fungible-asset)).

**Reply by ganymedio** [reply]
Thanks @kshitjee, good enough for me to mark as answered. 

Submitted a fix here https://github.com/aptos-labs/aptos-ts-sdk/pull/468

**Reply by kshitijc1506** [reply]
Thank you! Approved.
